File: main_code/poc_kmp_sdks/ios/Sources/ViewModels/HomeViewModel.swift
import Foundation
import SharedSDK

/// ViewModel for the Home view
/// Manages UI state and business logic for displaying users and products
@MainActor
class HomeViewModel: ObservableObject {
    @Published var users: [User] = []
    @Published var products: [Product] = []
    @Published var isLoading = false
    @Published var error: String?
    
    private let sharedSDK = SharedSDK()
    
    init() {
        // Initialize SDK with iOS-specific services
        let platformServices = IOSPlatformServices()
        // Note: In a real implementation, you would pass the platform services to the SDK
        sharedSDK.initialize()
    }
    
    /// Loads users and products from the shared SDK
    func loadData() {
        Task {
            isLoading = true
            error = nil
            
            do {
                // Load users and products concurrently
                async let usersResult = sharedSDK.getUsers()
                async let productsResult = sharedSDK.getProducts()
                
                let (fetchedUsers, fetchedProducts) = try await (usersResult, productsResult)
                
                users = fetchedUsers
                products = fetchedProducts
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }
    
    /// Searches users by name
    func searchUsers(query: String) {
        Task {
            do {
                let searchResults = try await sharedSDK.searchUsers(query: query)
                users = searchResults
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
    
    /// Searches products by title/description
    func searchProducts(query: String) {
        Task {
            do {
                let searchResults = try await sharedSDK.searchProducts(query: query)
                products = searchResults
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
    
    /// Clears the current error
    func clearError() {
        error = nil
    }
    
    /// Navigates to user details using the shared SDK
    func navigateToUserDetails(userId: String) {
        sharedSDK.navigateToUserDetails(userId: userId)
    }
    
    /// Navigates to product details using the shared SDK
    func navigateToProductDetails(productId: String) {
        sharedSDK.navigateToProductDetails(productId: productId)
    }
}

File: main_code/poc_kmp_sdks/android/src/androidMain/kotlin/ui/viewmodel/HomeViewModel.kt
package ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import main_code.SharedSDK
import main_code.data.Product
import main_code.data.User

/**
 * ViewModel for the Home screen
 * Manages UI state and business logic for displaying users and products
 */
class HomeViewModel(
    private val sharedSDK: SharedSDK = SharedSDK()
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()
    
    /**
     * Loads users and products from the shared SDK
     */
    fun loadData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            
            try {
                // Load users and products concurrently
                val users = sharedSDK.getUsers()
                val products = sharedSDK.getProducts()
                
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    users = users,
                    products = products
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Unknown error occurred"
                )
            }
        }
    }
    
    /**
     * Searches users by name
     */
    fun searchUsers(query: String) {
        viewModelScope.launch {
            try {
                val users = sharedSDK.searchUsers(query)
                _uiState.value = _uiState.value.copy(users = users)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "Search failed"
                )
            }
        }
    }
    
    /**
     * Searches products by title/description
     */
    fun searchProducts(query: String) {
        viewModelScope.launch {
            try {
                val products = sharedSDK.searchProducts(query)
                _uiState.value = _uiState.value.copy(products = products)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "Search failed"
                )
            }
        }
    }
    
    /**
     * Navigates to user details using the shared SDK
     */
    fun navigateToUserDetails(userId: String) {
        sharedSDK.navigateToUserDetails(userId)
    }
    
    /**
     * Navigates to product details using the shared SDK
     */
    fun navigateToProductDetails(productId: String) {
        sharedSDK.navigateToProductDetails(productId)
    }
}

/**
 * UI state for the Home screen
 */
data class HomeUiState(
    val isLoading: Boolean = false,
    val users: List<User> = emptyList(),
    val products: List<Product> = emptyList(),
    val error: String? = null
)